silkworm/db/datastore/snapshots/segment/kv_segment_reader.hpp

```diff
@@ -236,7 +236,7 @@ class KVSegmentKeysReader {
         value_type& value() const {
             Decoder& base_key_decoder = *(it_->first);
             // dynamic_cast is safe because TKeyDecoder was used when creating the Iterator
-            auto& key_decoder = dynamic_cast<TKeyDecoder&>(base_key_decoder);
+            DecoderConcept auto& key_decoder = dynamic_cast<TKeyDecoder&>(base_key_decoder);
```

Comment:
It feels that 2 CLion checks contradict each other.

The reason why it is auto, and not `TKeyDecoder&` is because some other CLion check (or maybe clang-tidy?) wants us to not repeat the type when dynamic/static cast is used, but now it forces to do extra specifications on auto, which simply repeats what was already specified at the TKeyDecoder declaration. So overspecification here contradicts underspecification (auto) from another check.

I'd keep the code as is, and disable this inspection locally in CLion.
Can keep this if you prefer, but for consistency please modify the segment_reader.hpp as well (non-kv).

Reply:
There are two different checks that come into play here: the one you mention is related to types (`TKeyDecoder` is a type), while the other I'm correcting here is related to concepts (`DecoderConcept` is a concept).

I'm not explicitly adding the type for `key_decoder` (which is still declared as `auto&`) but I'm adding a constraint over the automatically deduced type: this is called `constrained auto` and was introduced with C++20, see for example: [untangling-modern-c-constrained-auto-abbreviated-function-templates](https://medium.com/@rogerbooth/untangling-modern-c-constrained-auto-abbreviated-function-templates-d021efe75c0b) or [abbreviated-function-templates-and-constrained-auto](https://devblogs.microsoft.com/cppblog/abbreviated-function-templates-and-constrained-auto)

Reply:
Nevertheless, I agree that in this context it is redundant because we already have a clear type indication (`dynamic_cast<TKeyDecoder&>`), so I'm gonna undo the changes and disable this inspection locally in CLion.